### DISCUSSION TITLE
Permissions on Linux for pnet_datalink

### DISCUSSION THREAD

**Discussion by vinicius-smartme** [libpnet/libpnet#589] [General]
Hello.

I am using the `pnet_datalink` to create a simple program to send ethernet packets through a virtual interface, as follow:
- create a virtual interface using the command `sudo /sbin/ip li add veth0 type veth peer name veth1`;
- run a program which instantiates a `pnet_datalink::channel` to the virtual ethernet interface.

However, the Linux requires root permissions or the `CAP_NET_RAW` capability in order to instantiate a raw socket.

There are ways to solve this, like providing the `CAP_NET_RAW` to the executable. However, if I am modifying and running the program with `cargo run` or `cargo test` or when I am running it on a CI workflow, this won't help.

How could I properly provide the permissions so the program can run anyhow (executable, cargo, CI, etc)?

**Comment by stappersg** [comment]
On Wed, Oct 12, 2022 at 11:24:15PM -0700, vinicius-smartme wrote:
> 
> However, the Linux requires root permissions or the `CAP_NET_RAW`
> capability in order to instantiate a raw socket.
> 
> There are ways to solve this, like providing the `CAP_NET_RAW` to the
> executable. However, if I am modifying and running the program with
> `cargo run` or `cargo test` or when I am running it on a CI workflow,
> this won't help.
> 
> How could I properly provide the permissions so the program can run
> anyhow (executable, cargo, CI, etc)?
> 

For executable: https://man7.org/linux/man-pages/man8/setcap.8.html

For cargo, CI and etcetra:  Try to understand that your request
> How could I properly provide the permissions so the program ...
is about how to weak reasonable security.


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Reply by vinicius-smartme** [reply]
Thank you for your reply, @stappersg.

My idea for the permissions regarding the CI and cargo are basically for testing. I know it will weak the security, but those permissions wouldn't be used beyond testing.
